### PR TITLE
feat(dashboards): "Add Insight" flow

### PR DESCRIPTION
**Pull request by Twixes**
## Problem

In the dashboard view there's this button:

Screen Shot 2022-03-15 at 19 12 19

The issue is, it creates a new insight, but that insight ISN'T automatically assigned to the dashboard where the button was clicked, resulting in extra friction.
This is a [Quality Launch Week](https://github.com/PostHog/posthog/projects/15) problem.

## Changes

Now the button is called "Add Insight" and clicking on it takes you to a new insight that is assigned to the dashboard you came from with no extra interaction.

### CODE REVIEW

**Review by clarkus — COMMENTED** [review]
Is this always in a creation context? Is there a case for browsing for an existing insight to add it? We still have the 1:many dashboard:insight constraint, so that might be a factor - only browsing insights that aren't already on a dashboard.

**Comment by Twixes** [comment]
Here's the enhanced experience:

**Comment by clarkus** [comment]
Given the request originates on the dashboard, once added should the user be returned to their dashboard? That would be much stronger confirmation that the insight was added. It would also make it easier for that user to adjust the layout immediately and place that insight where they want it.

**Comment by Twixes** [comment]
That's a good point, updated:

**Comment by clarkus** [comment]
I think this is a good start but I have a few more questions about some possible edge cases. 

- When the origin of the request is from a dashboard, on insight creation we have an "adding to dashboard" button. Is this still necessary in this context? Should it name the dashboard we're adding the insight to? If I click it, can I break out of the flow and choose another dashboard? My point is that it might be doing more to confuse than direct towards a simpler path. What if in this scenario, the save action encompassed the dashboard add, too? "Save and add to dashboard".  This needs to be accounted for in https://github.com/PostHog/posthog/issues/9014. 
- Should there be an additional toast confirming the item was added to the dashboard? I see the insight saved toast, but that was it. 
- Did you consider the other comments about browsing for existing insights in this context? Should that be in scope?

**Comment by Twixes** [comment]
Good questions:
- The source dashboard's name is shown on hover, but not outright, to avoid it taking up a lot of space. You can't break out of the flow from the UI. I think "Save and add to dashboard" could indeed be clearer, however it'd break up the pattern we have with the "Add to dashboard/On dashboard" button – that's why I adapted the existing dashboard assignment button instead. Perhaps "Add to dashboard/On dashboard" needs rethinking too though.
- I suppose that "Insight saved" toast could just say "Insight saved to dashboard ${dashboard.name}" if it's assigned to a dashboard, WDYT?
- Hmm, I think it'd be nice to add existing insights from a list BUT that would expand the scope by a LOT. There are potentially very many insights, so this would require using the API's insight search capabilities. I'm not sure if this improvement is worth the effort.

**Comment by clarkus** [comment]
> The source dashboard's name is shown on hover, but not outright, to avoid it taking up a lot of space. You can't break out of the flow from the UI. I think "Save and add to dashboard" could indeed be clearer, however it'd break up the pattern we have with the "Add to dashboard/On dashboard" button – that's why I adapted the existing dashboard assignment button instead. Perhaps "Add to dashboard/On dashboard" needs rethinking too though.

I think we need to solve for this along with https://github.com/PostHog/posthog/issues/9014

> I suppose that "Insight saved" toast could just say "Insight saved to dashboard ${dashboard.name}" if it's assigned to a dashboard, WDYT?

Sounds good to me - the job to be done is confirming that it was added to the dashboard you started from. 

> Hmm, I think it'd be nice to add existing insights from a list BUT that would expand the scope by a LOT. There are potentially very many insights, so this would require using the API's insight search capabilities. I'm not sure if this improvement is worth the effort.

Maybe something for longer term but I do think users will want to reference something existing once we're able to support the 1:many relationship between insights and dashboards.

**Comment by clarkus** [comment]
I posted another variant of the save button at https://www.figma.com/file/gQBj9YnNgD8YW4nBwCVLZf/PostHog-App?node-id=8352%3A82865. It's not so much a new mode of saving (to users at least), it's more a "save this", but "then also do this other thing". So we're saving the insight as we would any other time, but along with that we're going to add a reference to a dashboard. Naming the dashboard affected would be good, but I realize there are constraints showing that.

**Comment by Twixes** [comment]
Here's the updated flow @clarkus 

### "New insight" from dashboard
Dashboard flow

### "New insight" individually
Individual flow

File: frontend/src/lib/components/LemonButton/LemonButton.scss [hunk 1]
```diff
@@ -86,6 +86,10 @@
     }
 }
 
+.LemonButton--translucent {
+    background: transparent;
+}
```
Comment (mariusandra) [inline frontend/src/lib/components/LemonButton/LemonButton.scss:91]: Always good to add new states to storybook.

File: frontend/src/scenes/insights/insightLogic.ts [hunk 2]
```diff
@@ -217,7 +219,7 @@ export const insightLogic = kea<insightLogicType>({
                     breakpoint()
 
                     // only update the fields that we changed
-                    const updatedInsight: Partial<InsightModel> = { ...values.insight }
+                    const updatedInsight = { ...values.insight } as InsightModel
```
Comment (mariusandra) [inline frontend/src/scenes/insights/insightLogic.ts:222]: This `as` smells fishy...

File: frontend/src/scenes/insights/insightLogic.ts [hunk 3]
```diff
@@ -438,6 +440,12 @@ export const insightLogic = kea<insightLogicType>({
                 setTagLoading: (_, { tagLoading }) => tagLoading,
             },
         ],
+        sourceDashboardId: [
+            null as number | null,
+            {
+                setSourceDashboardId: (_, { dashboardId }) => dashboardId,
+            },
+        ],
```
Comment (mariusandra) [inline frontend/src/scenes/insights/insightLogic.ts:448]: Conceptually this and actually the entire `saveInsight` feels like they make more sense on `insightSceneLogic`... 🤔

**Review by mariusandra — APPROVED** [review]
LGTM though with some comments. I'm working on something very closely related, but will pause now for a few hours: https://github.com/PostHog/posthog/pull/9143/files

**Review by clarkus — APPROVED** [review]
Looks good to me. The one thing I see is the button label for the "+ New insight" button on the dashboard. This should be "add insight" not "new insight" given that we're setting the context of adding a new reference to the dashboard.

**Comment by Twixes** [comment]
:+1: I was wavering between "Add" and "New" after I took in your point @clarkus on potentially allowing selection of an existing insight. In this case, "Add" does sound suggest that a little bit. But I definitely see the point that even with a new insight the paradigm is different, so updating to "Add".

**Comment by mariusandra** [comment]
merged to get ahead of conflicts in https://github.com/PostHog/posthog/pull/9143

**Pull request merged**